Add support for "SET SESSION sql_mode=' '"

javsanpar opened the pull request:
Related to issue #3863 and PR #4131

javsanpar committed:
Revert "Support sql_mode with spaces"

This reverts commit 0b75166cbfda4bb7ba583245fd8f729836ca70a5.

javsanpar committed:
Handle "SET SESSION sql_mode=' '" in handler_special_queries()

javsanpar committed:
Revert "Add test for sql_mode with spaces"

This reverts commit 76430156570adc09d387c248cae365528c0cc6e6.

javsanpar commented:
retest this please

JavierJF commented on lib/MySQL_Session.cpp:
You can perform the sum after the casting like this, removing the double cast:

```
static_cast<unsigned char*>(pkt_2.ptr) + 5
```

JavierJF commented on lib/MySQL_Session.cpp:
```diff
@@ -1536,6 +1536,34 @@ bool MySQL_Session::handler_special_queries(PtrSize_t *pkt) {
 	}
+	// Handle SQL_MODE with space. Issue #3863.
+	if (
+		pkt->size == strlen((char *)"SET SESSION sql_mode=' '") + 5 &&
```
Since used twice, I think we can precompute this `strlen((char *)"SET SESSION sql_mode=' '")`, compiler will do but, but to remove duplication.

JavierJF commented on lib/MySQL_Session.cpp:
```diff
@@ -1536,6 +1536,34 @@ bool MySQL_Session::handler_special_queries(PtrSize_t *pkt) {
+	// Handle SQL_MODE with space. Issue #3863.
+	if (
+		pkt->size == strlen((char *)"SET SESSION sql_mode=' '") + 5 &&
+		strncmp((char *)"SET SESSION sql_mode=' '", (char *)pkt->ptr+5, pkt->size-5) == 0
```
I don't know if this is a requirement (since this is a temporal corner case fix), but if not, we can be more loose here, using `strncasecmp` instead.

JavierJF commented on lib/MySQL_Session.cpp:
Nice catch on the offsets :+1: 

I'm thinking, since we already faced this situation let's fix it at API level so it shouldn't take place anymore, I propose to add a function to `QueryInfo` with following interface:

```
void Query_Info::replace_query_ptr(const PtrSize_t& pkt);
```

This is still unsafe because we would need were to call this function, but I feel like it's an small attempt of improvement. Also, for following better practices, we have a lot of hardcoded offsets in the code, I feel like we could also improve that with more self explanatory constants. All together:

```
void Query_Info::replace_query_ptr(const PtrSize_t& pkt) {
	QueryPointer = static_cast<unsigned char*>(pkt.ptr) + (NET_HEADER_SIZE + 1);
	QueryLength = pkt.size - (NET_HEADER_SIZE + 1);
}
```

We have also used `sizeof(mysql_hdr)` many times in the past, so, choose the one you like more. What do you think? If you like it, please peel free to replace the previous patterns in the code with the use of this safer variant, so there is no possible source of confusion in the future.

javsanpar committed:
Ignore case in SQL_MODE special query handler

javsanpar committed:
Add define for SQL_MODE special query

javsanpar committed:
Add helper function to replace query pointer

This function should avoid offset errors when replacing the query
pointer.

javsanpar committed:
Use replace_query_ptr() when needed

javsanpar committed:
Minor improvements when handling "SET SESSION sql_mode=' '"

To improve clarity, store the query in a variable an store its size in
other variable with "strlen()".

To improve homogeneity with libmariadbclient, rename define
NET_COMMAND_BYTE to NET_COMMAND_SIZE.

renecannao commented:
retest this please

javsanpar commented:
retest this please

mirostauder commented:
retest this please

mirostauder commented:
retest this please

mirostauder commented:
Can one of the admins verify this patch?

renecannao commented:
Closing this PR as obsolete. This functionality for supporting  was already implemented and merged via PR #4131 on March 2, 2023. This PR (#4150) was created on March 13, 2023 - 11 days after PR #4131 was already merged - making it duplicate work of an already-implemented feature. Additionally, this PR has merge conflicts and is no longer mergeable. Thank you @javsanpar for the contribution\!
